### PR TITLE
CLOUDP-111260: Allow mongocli to use tokens for authentication

### PR DESCRIPTION
**Pull request by gssbzn**
## Proposed changes

_Jira ticket:_ CLOUDP-111260

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](e2e/E2E-TESTS.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

```bash
./bin/mongocli auth login --profile dev

First, copy your one-time code: PW79-F9DX

Next, sign with your browser and enter the code.

Or go to https://account.mongodb.com/account/connect

Your code will expire after 10 minutes.
Successfully logged in.
Press Enter to continue your profile configuration
? Choose a default organization: 1 (5e7cdd07d59e2b4b392bcb9c)
? Choose a default project: 1 (61aa6fc08397d91def2ff012)
? Default Output Format:
? Default MongoDB Shell Path: /usr/local/bin/mongosh

Your profile is now configured.
To use this profile, you must set the flag [-P dev] for every command.
You can use [mongocli config set] to change these settings at a later time.
```

### CODE REVIEW

File: internal/cli/auth/login.go [hunk 1]
```diff
@@ -139,7 +206,8 @@ func LoginBuilder() *cobra.Command {
 	cmd.Flags().BoolVar(&opts.isGov, "gov", false, "Log in to Atlas for Government.")
 	cmd.Flags().BoolVar(&opts.isCloudManager, "cm", false, "Log in to Cloud Manager.")
 	cmd.Flags().BoolVar(&opts.noBrowser, "noBrowser", false, "Don't try to open a browser session.")
-
+	cmd.Flags().BoolVar(&opts.noProfile, "loginOnly", false, "Skip profile configuration.")
```
Comment (blva) [inline internal/cli/auth/login.go:209]: [nit] why not using similar names in `noProfile` and `loginOnly`?

File: internal/cli/auth/login.go [hunk 2]
```diff
@@ -110,10 +117,67 @@ Your code will expire after %.0f minutes.
...
+	var projectID string
+	if err != nil || len(pSlice) == 0 {
+		return errors.New("no projects")
+	}
+
+	p := prompt.NewProjectSelect(pSlice)
+	if err := survey.AskOne(p, &projectID); err != nil {
+		return err
+	}
+	opts.ProjectID = pMap[projectID]
```
Comment (tibulca) [inline internal/cli/auth/login.go:180]: most of the code here can be found also in [config command](https://github.com/mongodb/mongocli/blob/master/internal/cli/config/config.go)
For consistency (especially for the future, if we'll change any of these 2 commands) is it possible to share the common logic (I'd say config.go seems like a more suitable place), or we're trying to keep the commands as isolated as possible?

File: internal/cli/auth/login.go [hunk 3]
```diff
@@ -110,10 +117,67 @@ Your code will expire after %.0f minutes.
...
+		return err
+	}
+
+	opts.SetUpProject()
+	opts.SetUpOrg()
+	if err := survey.Ask(opts.DefaultQuestions(), opts); err != nil {
+		return err
+	}
+	opts.SetUpOutput()
+	opts.SetUpMongoSHPath()
```
Comment (blva) [inline internal/cli/auth/login.go:142]: could this be a function to make simplify `Run`?

File: internal/cli/auth/login.go [hunk 4]
```diff
@@ -110,10 +117,67 @@ Your code will expire after %.0f minutes.
...
 	if err := opts.config.Save(); err != nil {
 		return err
 	}
-	_, _ = fmt.Fprintf(opts.OutWriter, "Successfully logged in\n")
+	_, _ = fmt.Fprintf(opts.OutWriter, "\nYour profile is now configured.\n")
```
Comment (blva) [inline internal/cli/auth/login.go:146]: [q] I'm wondering if given that we are printing so many things here in this package, do we want to maybe create a function that simplifies this output with a void function like `print( "\nYour profile is now configured.\n")` 

and 
```
(opts *loginOpts) print(txt string) {
	_, _ = fmt.Fprintf(opts.OutWriter, txt)
}
```

or is this not very go-ish?

**Review by blva — COMMENTED** [review]
added minor comments, checking tests

File: internal/cli/auth/login.go (hunk 4 above)
Comment (gssbzn) [reply internal/cli/auth/login.go:146]: not sure if not goish but I gave it a try based on your comment, I decided against it as linters and the IDE loose the ability to flag when you make a formatting error

I intentionally removed one of the params and I no longer get warnings for it

Screenshot 2022-02-02 at 12 14 23

also given the varadic nature of printf wrapping it adds a bit of memory overhead to pack and unpack the args

File: internal/cli/auth/login.go (hunk 3 above)
Comment (gssbzn) [reply internal/cli/auth/login.go:142]: having a hard time to come up with a name for this but we could move the oauth part to a function which probably archives the same goal

File: internal/cli/auth/login.go (hunk 2 above)
Comment (gssbzn) [reply internal/cli/auth/login.go:180]: there's a main difference is that for config we have a fallback case where we don't do a select but let users input the id directly, this happens if they don't provide API keys which is allowed and would not happen for login, the main part that's shared `opts.Projects()` is being extracted as par of this PR

File: internal/cli/auth/login.go (hunk 1 above)
Comment (gssbzn) [reply internal/cli/auth/login.go]: it started as `noProfile` but then `loginOnly` felt as a better name, renamed the prop now

**Review by tibulca — APPROVED** [review]

File: internal/cli/auth/login.go (hunk 4 above)
Comment (blva) [reply internal/cli/auth/login.go]: nice check! fair enough

**Review by blva — APPROVED** [review]
LGTM

**Pull request merged**